Como podemos usar o poder de Deep Learning para realizar tarefas de NLP? Bom, podemos pensar em aplicações mais complexas como _Speech Recognition_; _Chatbots_ que entendem perguntas complexas, entre outras situações. Contudo, podemos tirar proveito de NLP com Deep Learning de uma forma "relativamente" (áspas intencionais) simples: que é através do _Word2Vec_.

O [_Word2Vec_](https://code.google.com/archive/p/word2vec/) é uma aplicação desenvolvida pela Google por volta de 2013 que nada mais é do que um modelo de _Embedding de palavras_. Para aqueles que queiram uma definição mais formal, essa [apresentação](https://docs.google.com/file/d/0B7XkCwpI5KDYRWRnd1RzWXQ2TWc/edit) pode ser util.

## Embeddings de Palavras

Quando lidamos com palavras em texto, podemos ter uma situação em que temos milhares de classes para prever, uma para cada palavra. Uma abordagem que podemos usar é a conhecida como **one hot encoding**, em que o número de neurônios da nossa Rede Neural equivale ao número de palavras do nosso _dataset_ e ao entrar diferentes palavras, sua posição no vocabulário é setada como 1 e nos demais, 0. Contudo, esse tipo de situação é extreamente ineficiente, já que por exemplo, podemos ter uma situação com apenas um elemento setado como 1 e 50,000 como 0. Computacionalmente é um processo caro.

![One Hot Encoding](img/one-hot.png)

Para resolver esse problema nós usamos o que pode ser chamado de _Word Embedding_._Embeddings_ nada mais são do que uma camada oculta (a camada de _Embedding_) de uma Rede Neural Artificial como essa que a gente acabou de ver. Nos _Embeddings_, não é preciso computar todos os neurônios, mas apenas pegar a linha dos vetores de peso (_embedding weights_). A gente pode fazer isso porquê a multiplicação de um vetor _one hot encoding_ por essa camada irá resultar apenas no valor da linha correspondente ao 1.

![One Hot Multiply](img/one-multiply.png)


Logo, quando tratamos com _Embeddings_, computacionalmente falando não ocorre multiplicação matricial, mas apenas um _lookup_. Lembra que os computadores entendem apenas números? Então, suponha que a palavra _house_ tenha sido encodada na linha 958 enquanto _life_ na 1056. Para obter os valores de peso de _house_ basta olhar na linha 958 da tabela de pesos. Esse processo é conhecido como **_embedding lookup_** e o número de neurônios dessa camada oculta é conhecida como **_embedding dimension_**.

Embeddings não são usados para palavras, apenas. Você pode usá-los para qualquer modelo que tenha um número massivo de classe. Um desses modelos é o _Word2Vec_, que usa modelos de palavras para encontrar representações semânticas das palavras.

## A intuição do Word2Vec

O _Word2Vec_ foi uma aplicação capaz de mostrar que redes neurais conseguem encontrar assimilações entre palavras. Por exemplo:
- **homem** está para **mulher** assim como **rei** está para?

### Para pensar um pouco:

Em seu paper original, Mikolov usou o dataset do _Google News_ que era composto de 3 milhões de palavras e frases, resultando em mais de 100 bilhões de palavras. Nele, foi possível encontrar associações bem interessantes, como:

- Microsoft - Windows ; Google- ?; IBM- ?; Apple -?

![Word2Vec](img/word2vec.png)

### Por debaixo dos panos

![SemanticSpace](img/semanticspace.png)


Sendo assim, é possível encontrar relações bem interessantes, não só limitadas a relações pessoais, como também de verbos e lugares:


![Relationships](img/relationships.png)


### A arquitetura por trás disso

A ideia aqui não é ensinar a como implementar o _Word2Vec_ do zero, mas sim entender toda a intuição por trás dele. Lembra que mais cedo eu falei que estruturas de _Embedding_ nada mais são do que camadas ocultas de Redes Neurais? Então, as duas arquiteturas que implementam o _Word2Vec_ são **Redes Neurais**: o continuous Bag of Words (**CBOW**) e o **Skip-gram**.

![Architetures](img/arc.png)

Como podemos ver, as diferenças entre cada uma das arquiteturas é a seguinte:

-**CBOW**: Dado um contexto (uma janela de palavras), eu tento prever a palavra do meio da janela.

-**Skip-gram**: Dada uma palavra, eu tento prever as palavras em volta (ou o contexto) em que ela está inserida.

Pelo paper, a arquitetura _skipgram_ tende a performar melhor que o _CBOW_.


## Preparando os dados

Para a aula de hoje, usaremos apenas o pacote de processamento de texto `gensim`. A ideia é mostrar que ele é uma alternativa ao `scikit-learn` para algumas etapas de pré processamento, mas também é válido dizer que ambos são complementares. Em outras palavras, é possível usar a saída de algum processo do `gensim` no `scikit` e vice versa.

Uma coisa legal do Word2Vec é que ele pode ser considerado como um tipo de aprendizado **não** surpervisionado. Em outras palavras, ele é capaz de aprender de textos que não possuam um rótulo (ou label). Para ilustrar isso, para o treinamento aqui, além de usarmos o _labeledTrainData.tsv_, também usaremos o _unlabeledTrainData.tsv_, que contém 50 mil reviews adicionais, sem nenhuma label.

In [ ]:
# Faca a leitura dos arquivos labeledTrainData.tsv;testData.tsv e unlabeledTrainData.tsv
#Armazene seus valores nas variáveis train, test e unlabeled_train respectivamente

#Lembre-se de importar o pandas !
train = 
test =
unlabeled_train = 

In [ ]:
train.shape

In [3]:
test.shape

(25000, 2)

In [4]:
unlabeled_train.shape

(50000, 2)

## Pré Processando o texto

É interessante realizarmos o pré processamento dos textos que vamos trabalhar, igual fizemos na aula passada. Contudo, diferente do modelo **BoW** que vimos, o _Word2Vec_ **leva** em consideração o contexto das palavras. Sendo assim, talvez não seja muito legal removermos as *stop words*, assim como reduzir as palavras ao seu radical (stemming ou lemma) pode não ser desejável.

Contudo, um pré processamento interessante é remover as URL's, uma vez que a chance de um site ser citado em um review é baixo e não estamos muito interessados na relação entre sites, mas no contexto em que levou a pessoa a citar alguma fonte.

In [20]:
# Import various modules for string cleaning
from bs4 import BeautifulSoup
import re
from nltk.corpus import stopwords


def review_to_wordlist( review, remove_stopwords=False ):
    # Function to convert a document to a sequence of words,
    # optionally removing stop words.  Returns a list of words.
    #
    
    # 1. Ignore URLs
    review_text = re.sub("IMPLEMENTE AQUI A REGEX QUE IGNORA URL","<URL>", review)
    # 2. Remove HTML
    review_text = #IMPLEMENTE AQUI A FUNCAO QUE REMOVE OS <br> E OUTRAS COISAS DO HTML
    #  
    # 3. Remove non-letters
    # IMPLEMENTE AQUI A FUNCAO QUE REMOVE CARACTER ESPECIAL
    #
    # 4. Convert words to lower case and split them
    words = #IMPLEMENTE AQUI A FUNCAO QUE REMOVE EXCESSO DE ESPACOS, QUE CONVERTE TUDO
            #PARA MINUSCULA E SPLITA AS PALAVRAS
    #
    # 5. Optionally remove stop words (false by default)
    #INCLUA AQUI A PARTE DE REMOVER STOP WORDS
    #LEMBRE-SE QUE AGORA ELA EH UMA BOOLEAN E NAO DEVE SER EXECUTADA
    #SE SEU VALOR FOR FALSO
    #
    # 6. Return a list of words
    return(words)

Uma vez limpa, agora precisamos converter os nossos dados para o tipo de dado que a biblioteca do `gensim` aceite. O _Word2Vec_ recebe como entrada sentenças únicas, cada uma formada por uma lista de palavras. Ou seja, precisamos de uma lista de listas.

Definir onde começa ou onde acaba uma sentença não é algo muito claro. Existem sentenças que podem terminar com "?";".";"!" entre outras. Por conta disso, usaremos o pacote sent_tokenize do NLTK para fazer esse tipo de split pra gente. 


In [21]:
from nltk.tokenize import sent_tokenize
from tqdm import tqdm

In [22]:
sentence = 'O André roubou pão na casa do João ! Quem, Eu? Você ! Eu não. Então quem foi?'

In [23]:
sent_tokenize(sentence)

['O André roubou pão na casa do João !',
 'Quem, Eu?',
 'Você !',
 'Eu não.',
 'Então quem foi?']

In [24]:
# Define a function to split a review into parsed sentences
def review_to_sentences( review,remove_stopwords=False ):
    # Function to split a review into parsed sentences. Returns a 
    # list of sentences, where each sentence is a list of words

In [25]:
review_to_sentences(sentence)

[['o', 'andré', 'roubou', 'pão', 'na', 'casa', 'do', 'joão'],
 ['quem', 'eu'],
 ['você'],
 ['eu', 'não'],
 ['então', 'quem', 'foi']]

In [8]:
sentences = []  # Inicialize uma lista de sentenca e inclua nela 
                #todos os reviews processados


  1%|          | 218/25000 [00:04<09:20, 44.21it/s]/Users/abarbosa/miniconda3/envs/somostera/lib/python3.6/site-packages/bs4/__init__.py:219: UserWarning: "b'.'" looks like a filename, not markup. You should probably open this file and pass the filehandle into Beautiful Soup.
  ' Beautiful Soup.' % markup)
  0%|          | 10/50000 [00:00<08:32, 97.61it/s]

Starting Unlabeled Sentences


 62%|██████▏   | 31189/50000 [09:00<06:18, 49.71it/s]/Users/abarbosa/miniconda3/envs/somostera/lib/python3.6/site-packages/bs4/__init__.py:219: UserWarning: "b'..'" looks like a filename, not markup. You should probably open this file and pass the filehandle into Beautiful Soup.
  ' Beautiful Soup.' % markup)
100%|██████████| 50000/50000 [14:25<00:00, 57.80it/s]


In [27]:
array = [1,2,3,4]
brray = [5,6,7,8]

array.append(brray)
array

[1, 2, 3, 4, [5, 6, 7, 8]]

In [31]:
array = [1,2,3,4]
array += brray
array

[1, 2, 3, 4, 5, 6, 7, 8]

In [32]:
array = [1,2,3,4]
array.extend(brray)
array

[1, 2, 3, 4, 5, 6, 7, 8]

In [9]:
len(sentences)

795538

In [10]:
sentences[0]

['with',
 'all',
 'this',
 'stuff',
 'going',
 'down',
 'at',
 'the',
 'moment',
 'with',
 'mj',
 'i',
 've',
 'started',
 'listening',
 'to',
 'his',
 'music',
 'watching',
 'the',
 'odd',
 'documentary',
 'here',
 'and',
 'there',
 'watched',
 'the',
 'wiz',
 'and',
 'watched',
 'moonwalker',
 'again']

In [11]:
sentences[1]

['maybe',
 'i',
 'just',
 'want',
 'to',
 'get',
 'a',
 'certain',
 'insight',
 'into',
 'this',
 'guy',
 'who',
 'i',
 'thought',
 'was',
 'really',
 'cool',
 'in',
 'the',
 'eighties',
 'just',
 'to',
 'maybe',
 'make',
 'up',
 'my',
 'mind',
 'whether',
 'he',
 'is',
 'guilty',
 'or',
 'innocent']

## Treinando o modelo Word2Vec

Com a lista de sentenças devidamente parseada, basta treinarmos o nosso modelo. Aqui tem uma lista dos parâmetros especificando o que cada uma coisa faz.

- **Architecture**: As opções de arquitetura são os algoritmos possíveis para a realização do Word2Vec. As opções aqui são o _skip-gram_ (default) e o _continuous bag of words_ (_cbow_).

- **Training algorithm**: _Hierarchical softmax_ (default) ou _negative sampling_. Aqui, o default funcionou bem.

- **Downsampling of frequent words**: A documentação do Google sugere valores entre .00001 and .001. Aqui, algo epor volta de 0.001 parece que melhorou o modelo final.

- **Word vector dimensionality**: Aqui a ideia de _feature_ é parecida com o modelo que vimos do _BoW_, mas cada coluna não necessariamente é uma palavra. Logo, mais _features_ podem resultar em tempos maiores de processamento, mas nem sempre em modelos ideiais. O default aqui é 300.

- **Context / window size**: Quantas janelas de contexto o algoritmo de treinamento deve levar em consideração. 10 parece funcionar bem para o softmax hierarquico (quanto mais melhor, até certo ponto).

- **Worker threads**: Numero de processos para correrem em paraleolo. Isso varia de máquina para máquina, mas algo entre 4 e 6 deve funcionar na maioria dos sistemas.

- **Minimum word count**: Isso ajuda a limitar o tamanho do vocabulario de palavras significativas. Qualquer palavra que não ocorre ao menos esse número de vezes considerando **todos** os documentos é ignorada. Valores razoáveis são entre 10 e 100. Nesse caso, dado que cada filme ocorre 30 vezes, o valor setado de _minimum word count_ é de 40. Quanto maior o vocabulário, maior o tempo de execução é impactado.

In [12]:
# Import the built-in logging module and configure it so that Word2Vec 
# creates nice output messages
import logging
logging.basicConfig(format='%(asctime)s : %(levelname)s : %(message)s',\
    level=logging.INFO)

# Set values for various parameters
num_features =     # Word vector dimensionality                      
min_word_count =    # Minimum word count                        
num_workers =        # Number of threads to run in parallel
context =           # Context window size                                                                                    
downsampling = 1e-3   # Downsample setting for frequent words

# Initialize and train the model (this will take some time)
from gensim.models import word2vec
print ("Training model...")
model = word2vec.Word2Vec(sentences, workers=, \
            size=, min_count = , \
            window = , sample = downsampling)

# If you don't plan to train the model any further, calling 
# init_sims will make the model much more memory-efficient.
model.init_sims(replace=True)

# It can be helpful to create a meaningful model name and 
# save the model for later use. You can load it later using Word2Vec.load()
model_name = "300features_40minwords_10context"
model.save(model_name)

2017-11-29 02:15:44,723 : INFO : 'pattern' package not found; tag filters are not available for English
2017-11-29 02:15:44,729 : INFO : collecting all words and their counts
2017-11-29 02:15:44,730 : INFO : PROGRESS: at sentence #0, processed 0 words, keeping 0 word types
2017-11-29 02:15:44,791 : INFO : PROGRESS: at sentence #10000, processed 227183 words, keeping 18050 word types
2017-11-29 02:15:44,857 : INFO : PROGRESS: at sentence #20000, processed 454403 words, keeping 25345 word types
2017-11-29 02:15:44,915 : INFO : PROGRESS: at sentence #30000, processed 675010 words, keeping 30508 word types


Training model...


2017-11-29 02:15:44,986 : INFO : PROGRESS: at sentence #40000, processed 902687 words, keeping 34903 word types
2017-11-29 02:15:45,051 : INFO : PROGRESS: at sentence #50000, processed 1123085 words, keeping 38385 word types
2017-11-29 02:15:45,119 : INFO : PROGRESS: at sentence #60000, processed 1345767 words, keeping 41405 word types
2017-11-29 02:15:45,181 : INFO : PROGRESS: at sentence #70000, processed 1570193 words, keeping 44067 word types
2017-11-29 02:15:45,245 : INFO : PROGRESS: at sentence #80000, processed 1790610 words, keeping 46494 word types
2017-11-29 02:15:45,312 : INFO : PROGRESS: at sentence #90000, processed 2016019 words, keeping 48972 word types
2017-11-29 02:15:45,381 : INFO : PROGRESS: at sentence #100000, processed 2239092 words, keeping 51099 word types
2017-11-29 02:15:45,445 : INFO : PROGRESS: at sentence #110000, processed 2460033 words, keeping 53028 word types
2017-11-29 02:15:45,512 : INFO : PROGRESS: at sentence #120000, processed 2683309 words, keepin

2017-11-29 02:15:50,035 : INFO : PROGRESS: at sentence #760000, processed 17083737 words, keeping 124327 word types
2017-11-29 02:15:50,106 : INFO : PROGRESS: at sentence #770000, processed 17312165 words, keeping 125128 word types
2017-11-29 02:15:50,173 : INFO : PROGRESS: at sentence #780000, processed 17543573 words, keeping 125864 word types
2017-11-29 02:15:50,242 : INFO : PROGRESS: at sentence #790000, processed 17771837 words, keeping 126562 word types
2017-11-29 02:15:50,281 : INFO : collected 127017 word types from a corpus of 17895591 raw words and 795538 sentences
2017-11-29 02:15:50,282 : INFO : Loading a fresh vocabulary
2017-11-29 02:15:50,377 : INFO : min_count=40 retains 16717 unique words (13% of original 127017, drops 110300)
2017-11-29 02:15:50,378 : INFO : min_count=40 leaves 17328137 word corpus (96% of original 17895591, drops 567454)
2017-11-29 02:15:50,432 : INFO : deleting the raw counts dictionary of 127017 items
2017-11-29 02:15:50,438 : INFO : sample=0.001 d

2017-11-29 02:16:58,152 : INFO : PROGRESS: at 99.79% examples, 953334 words/s, in_qsize 7, out_qsize 0
2017-11-29 02:16:58,290 : INFO : worker thread finished; awaiting finish of 3 more threads
2017-11-29 02:16:58,296 : INFO : worker thread finished; awaiting finish of 2 more threads
2017-11-29 02:16:58,301 : INFO : worker thread finished; awaiting finish of 1 more threads
2017-11-29 02:16:58,307 : INFO : worker thread finished; awaiting finish of 0 more threads
2017-11-29 02:16:58,307 : INFO : training on 89477955 raw words (64280920 effective words) took 67.4s, 953145 effective words/s
2017-11-29 02:16:58,309 : INFO : precomputing L2-norms of word weight vectors
2017-11-29 02:16:58,478 : INFO : saving Word2Vec object under 300features_40minwords_10context, separately None
2017-11-29 02:16:58,479 : INFO : not storing attribute syn0norm
2017-11-29 02:16:58,480 : INFO : not storing attribute cum_table
2017-11-29 02:16:59,127 : INFO : saved 300features_40minwords_10context


## Explorando os resultados

A função `doesnt_match` tentará deduzir qual palavra em um set é a mais _distinta_ das outras.



In [13]:
#apenas paises
model.doesnt_match("".split())

'berlin'

De forma análoga, a função `most_similar` tentará deduzir qual palavra em um set é a mais _semelhante_ das outras.

In [14]:
model.most_similar("")

#podemos ate ter um indicio para sentiment analysis
model.most_similar("")

[('terrible', 0.7730972766876221),
 ('horrible', 0.7526340484619141),
 ('atrocious', 0.7306756973266602),
 ('dreadful', 0.7170031666755676),
 ('abysmal', 0.7054095268249512),
 ('horrendous', 0.6861761808395386),
 ('horrid', 0.6846473217010498),
 ('appalling', 0.6617923974990845),
 ('lousy', 0.6321056485176086),
 ('crappy', 0.615153431892395)]




## Mas como definir similaridade? Que uma palavra é similar à outra?

A métrica mais famosa para isso é a _Similaridade de Coscenos_ ou _Cosine Similarity_.

Vamos voltar à definição de espaço semantico dada no começo da aula.

![Male-Female](img/male-female.png)

Olhando um pouco debaixo dos panos (porque na situação real é como se estivessemos olhando para um vetor de 300 dimensões, o que é muito abstrato), vamos projetar **apenas** essa situação de _Male-Female_


![Male-Female](img/sematic-male-female.png)


Ao olharmos esse espaço, vemos claramente que as palavras semânticamente próximas a _male_ também são próximas entre si. O mesmo acontece com _female_. Mas como que a gente encontra esse *valor*? Basta calcularmos o **cosceno** entre os vetores que formam essas palavras !

Lembrando um pouquinho de matemática da faculdade, dado o produto escalar entre dois vetores e suas normas eu consigo encontrar o cosceno formado entre os dois:

$A \cdot B = \lVert A \rVert \lVert B \rVert \cos\theta   $

![Cosine](img/cosine-similarity.png)


Em NLP essa métrica é muito comum quando queremos calcular a similaridade entre documentos.

## Entendendo a representação numérica das palavras

Agora nós treinamos um modelo que consegue entender a relação semântica entre as palavras. Para algumas aplicações é exatamente isso que a gente quer, mas o que podemos fazer com isso? Os vetores de palavras criados pelo nosso modelo foram armazenados em um array `numpy` chamado `syn0`

In [15]:
from gensim.models import Word2Vec 
model = Word2Vec.load("300features_40minwords_10context")
type(model.wv.syn0)

2017-11-29 02:16:59,201 : INFO : loading Word2Vec object from 300features_40minwords_10context
2017-11-29 02:16:59,553 : INFO : loading wv recursively from 300features_40minwords_10context.wv.* with mmap=None
2017-11-29 02:16:59,554 : INFO : setting ignored attribute syn0norm to None
2017-11-29 02:16:59,555 : INFO : setting ignored attribute cum_table to None
2017-11-29 02:16:59,556 : INFO : loaded 300features_40minwords_10context


numpy.ndarray

In [16]:
model.wv.syn0.shape

(16717, 300)

Lembre-se que a gente limitou o número de ocorrências de palavras para um mínimo de 40. Isso justifica a menor quantidade de vocabulário, sendo formado por 16717 palavras.

In [17]:
model["flower"]

array([  4.14542146e-02,   1.36064023e-01,  -4.50171567e-02,
         3.66699137e-02,   5.50786778e-03,  -4.65489067e-02,
         9.40594971e-02,   3.61240245e-02,  -1.84720531e-02,
        -4.95337918e-02,   4.64605819e-03,  -8.72984678e-02,
        -3.61285551e-04,   9.04739499e-02,  -5.85301267e-03,
         9.78970993e-03,  -8.90556071e-03,   6.74419105e-02,
         2.54531633e-02,  -1.24042602e-02,  -1.13968197e-02,
         1.45032071e-02,  -5.83921298e-02,  -9.96814389e-03,
         1.35538206e-01,   8.33753794e-02,  -6.16430538e-03,
         3.69862095e-02,   8.45942795e-02,   1.28843904e-01,
         3.15543227e-02,  -1.10337615e-01,  -6.18141070e-02,
        -9.54130739e-02,   9.34280083e-02,   4.09199670e-02,
        -4.39016633e-02,   2.57810298e-02,  -2.82172319e-02,
         6.09959327e-02,   1.52271669e-02,   3.31028625e-02,
        -9.84279960e-02,   5.35987169e-02,   9.68147144e-02,
         4.40609679e-02,  -4.52714823e-02,   6.65505044e-03,
         1.26678981e-02,

In [18]:
len(model["flower"])

300

## De palavras para reviews 

Um problema que temos ao lidar com o dataset do IMDB é que os reviews tem tamanho variado. Precisaríamos, então, encontrar uma forma de pegar esses vetores de palavras e transformá-los em um conjunto de features que tem o mesmo tamanho para o mesmo _review_.

Já que todas as palavras são representadas por um vetor de 300 dimensões, nós podemos usar operações vetoriais para combinar palavras. Por exemplo, podemos pegar todas as palavras que compõe um review e somá-las ou tirar a média. No caso, vamos tirar a média (contudo, aqui nós não precisamos mais do contexto das palavras e, então, podemos remover stopwords).

In [19]:
import numpy as np  # Make sure that numpy is imported

def makeFeatureVec(review, model, num_features):
    # Function to average all of the word vectors in a given
    # paragraph
    #
    # Pre-initialize an empty numpy array (for speed)
    featureVec = np.zeros((num_features,),dtype="float32")
    #
    nwords = 0.
    # 
    # Index2word is a list that contains the names of the words in 
    # the model's vocabulary. Convert it to a set, for speed 
    index2word_set = set(model.wv.index2word)
    #
    # Loop over each word in the review and, if it is in the model's
    # vocaublary, add its feature vector to the total
    #Hint: np.add should be useful
    
    # 
    # Divide the result by the number of words to get the average
    #Hint: use np.divide
    return featureVec


def getAvgFeatureVecs(reviews, model, num_features):
    # Given a set of reviews (each one a list of words), calculate 
    # the average feature vector for each one and return a 2D numpy array 
    # 
    # 
    # Preallocate a 2D numpy array, for speed
    reviewFeatureVecs = np.zeros((len(reviews),num_features),dtype="float32")
    # 
    # Loop through the reviews
    for review in tqdm(reviews,total=len(reviews)):

    #  Implement here the function that will calculate the average for each review     
    #

    return reviewFeatureVecs

In [20]:
# ****************************************************************
# Calculate average feature vectors for training and testing sets,
# using the functions we defined above. Notice that we now use stop word
# removal.

clean_train_reviews = []


trainDataVecs = 

print ("Creating average feature vecs for test reviews")
clean_test_reviews = []


testDataVecs = 

  0%|          | 56/25000 [00:00<00:45, 548.47it/s]

Creating average feature vecs for test reviews


100%|██████████| 25000/25000 [00:33<00:00, 746.82it/s]


Aqui, como a ideia é usar o vetor de features como treino, estamos lidando novamente com um aprendizado **surpervisionado** e, então, usaremos apenas os dados com _label_.

In [21]:
# Fit a random forest to the training data, using 100 trees
from sklearn.ensemble import RandomForestClassifier
forest = RandomForestClassifier( n_estimators = 100 )

forest = forest.fit( trainDataVecs, train["sentiment"] )

# Test & extract results 
result = forest.predict( testDataVecs )

# Write the test results 
output = pd.DataFrame( data={"id":test["id"],"review":test["review"],"sentiment":result} )

output

id                                             review  \
0      "12311_10"  "Naturally in a film who's main themes are of ...   
1        "8348_2"  "This movie is a disaster within a disaster fi...   
2        "5828_4"  "All in all, this is a movie for kids. We saw ...   
3        "7186_2"  "Afraid of the Dark left me with the impressio...   
4       "12128_7"  "A very accurate depiction of small time mob l...   
5        "2913_8"  "...as valuable as King Tut's tomb! (OK, maybe...   
6        "4396_1"  "This has to be one of the biggest misfires ev...   
7         "395_2"  "This is one of those movies I watched, and wo...   
8       "10616_1"  "The worst movie i've seen in years (and i've ...   
9        "9074_9"  "Five medical students (Kevin Bacon, David Lab...   
10       "9252_3"  "'The Mill on the Floss' was one of the lesser...   
11       "9896_9"  "I just saw this film at the phoenix film fest...   
12        "574_4"  "\"The Love Letter\" is one of those movies th...   
13      "11182_8"  "Another fantastic offering from the Monkey Is...   
14      "11656_4"  "This was included on the disk \"Shorts: Volum...   
15       "2322_4"  "I'm not really much of an Abbott & Costello f...   
16       "8703_1"  "This movie was dreadful. Biblically very inac...   
17       "7483_1"  "I don't think I've ever gave something a 1/10...   
18      "6007_10"  "Excellent story-telling and cinematography. P...   
19      "12424_4"  "I completely forgot that I'd seen this within...   
20       "4672_1"  "I like action movies. I have a softspot for \...   
21      "10841_3"  "This is one of the worst Sandra Bullock movie...   
22       "8954_7"  "Watched this flick on Saturday afternoon cabl...   
23       "7392_1"  "I went to see \"TKIA\" with high expectations...   
24      "10288_8"  "All credit to writer/director Gilles Mimouni ...   
25       "5343_4"  "As a writing teacher, there are two ending I ...   
26       "4950_1"  "I don't know why this has gotten any decent r...   
27       "9257_4"  "This film was released in the UK under the na...   
28       "8689_3"  "Uncle Fred Olen Ray once again gives us a lit...   
29       "4480_2"  "OK, it's watchable if you are sick in bed or ...   
...           ...                                                ...   
24970   "6857_10"  "With \"Anatomy\" the german film producers ha...   
24971   "11091_8"  "This movie is one of my all-time favorites. I...   
24972    "4167_2"  "I found Code 46 very disappointing. I thought...   
24973     "679_4"  "Tamara Anderson and her family are moving onc...   
24974   "10147_1"  "Now I've seen it all. Just when I thought it ...   
24975    "6875_1"  "In this movie everything possible was wrong a...   
24976    "923_10"  "Well every scene so perfectly presented. Neve...   
24977    "6200_8"  "Sleeper Cell is what 24 should have been. 24 ...   
24978    "7208_8"  "Not for everyone, but I really like it. Nice ...   
24979    "5363_8"  "Set just before the Second World War, this is...   
24980    "4067_8"  "Contains Spoiler The movie is a good action/c...   
24981    "1773_7"  "This is one of several period sea-faring yarn...   
24982   "1498_10"  "Hearkening back to those \"Good Old Days\" of...   
24983  "10497_10"  "I thought this to be a pretty good example of...   
24984   "3444_10"  "Seeing this film, or rather set of films, in ...   
24985     "588_2"  "I didn't like this movie for many reasons - V...   
24986    "9678_9"  "I absolutely love this show!!!!!!!, Its basic...   
24987    "1983_9"  "eXistenZ combines director David Cronenberg's...   
24988    "5012_3"  "this movie is allegedly a comedy.so where did...   
24989   "12240_2"  "The Comebacks is a spoof on inspirational spo...   
24990    "5071_2"  "I'd love to write a little summary of this mo...   
24991    "5078_2"  "Obvious tailored vehicle for Ryan Philippe. I...   
24992   "10069_3"  "<br /><br />JURASSIC PARK III *___ Adventure ...   
24993    "7407_8"  "If you're even mildly interested in

## Mas analisando o Word2Vec 

Os resultados foram melhores que o BoW?

Se analisarmos de uma forma minuciosa, vemos que os resultados do BoW foram **um pouco melhores** que o do Word2Vec, mas porquê isso ocorre?

O fato da gente utilizar a média para criar as sentenças faz com que nós percamos a ordem das palavras (além de perdermos informação ao resumir os dados por meio da média), fazendo com que a abordagem seja muito parecida com o conceito do Bag Of Words, de forma que os métodos até sejam considerados equivalentes.

Algumas coisas para melhorar:
- Dados, dados dados. Lembre-se que o paper original usou um corpus de mais de um **bilhão** de palavras enquanto os nossos dados tinham aproximadamente 18 **milhões** de palavras. É possível usar vetorer pré treinados para usar em nossos dados e talvez isso resulte em resultados mais interessantes.
